File: cpp-gtest/RomanNumeral.cpp
#include <cstring>
#include "RomanNumeral.h"

static RomanDigit singleDigit[] = {
        {.value = 1, .symbol = "I"},
        {.value = 4, .symbol = "IV"},
        {.value = 5, .symbol = "V"},
        {.value = 9, .symbol = "IX"},
        {.value = 10, .symbol = "X"},
        {.value = 40, .symbol = "XL"},
        {.value = 50, .symbol = "L"}
};

string RomanNumeral::toRoman(int value) {
    string result = ""; // REDUNDANT
    int index = sizeof(singleDigit)/sizeof(RomanDigit) - 1;
    RomanDigit currentDigit = singleDigit[index];

    while(value > 0) {
        if (currentDigit.value <= value) {
            value -= currentDigit.value;
            result += currentDigit.symbol;
        } else {
            currentDigit = singleDigit[--index];
        }
    }
    return result;
}

File: java-junit/src/main/java/RomanNumeral.java
/**
 * Created by cdowling on 7/9/16.
 */
public class RomanNumeral {

    private static RomanDigit [] singleDigits = new RomanDigit[] {
            new RomanDigit(1, "I"),
            new RomanDigit(5, "V"),
            new RomanDigit(10, "X")
    };

    private RomanDigit findSingleDigit(int digit) {
        for(RomanDigit roman : singleDigits ) {
            if (roman.value == digit) {
                return roman;
            }
        }
        return null;
    }

    public String toRoman(int value) {
        RomanDigit digit = findSingleDigit(value);
        String result = "";
        if (digit != null) {
            result = result + digit.symbol;
        } else {


            switch (value) {
                case 2:
                    result = "II";
                    break;
                case 7:
                    result =  "VII";
                    break;
                default:
                    result = "undefined";
            }
        }
        return result;
    }
}
